src/frontends/onnx/frontend/src/core/tensor.hpp

```diff
@@ -260,8 +270,22 @@ class Tensor {
         std::shared_ptr<ov::op::v0::Constant> constant{nullptr};
         auto data = get_data<T>();
         auto data_size = data.size();
+        switch (type) {
+        case ov::element::i4:
+        case ov::element::u4:
+            data_size *= 2;  // Each byte contains 2 data items
+            break;
+        }
         if (data_size == shape_size(m_shape)) {
-            constant = std::make_shared<ov::op::v0::Constant>(type, m_shape, data);
+            switch (type) {
+            case ov::element::i4:
+            case ov::element::u4:
+                constant = std::make_shared<ov::op::v0::Constant>(type, m_shape, data.data());
+                break;
+            default:
+                constant = std::make_shared<ov::op::v0::Constant>(type, m_shape, data);
+                break;
+            }
```

Comment:
```suggestion
            constant = std::make_shared<ov::op::v0::Constant>(type, m_shape, data.data());
```
As number elements in `data` is same as in Constant it should work in all cases.